spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapter.java

```diff
@@ -145,24 +145,26 @@ public void nack() {
 			});
 		}
 
-		boolean messageNacked = false;
-
 		try {
 			sendMessage(getMessageBuilderFactory()
 					.withPayload(message.getPayload())
 					.copyHeaders(messageHeaders)
 					.build());
+
+			if (this.ackMode == AckMode.AUTO_ACK || this.ackMode == AckMode.AUTO) {
+				message.ack();
```

Comment:
Why should we move `ack()` outside of the finally?
Can we just fix the bug that was introduced in the last PR without further changes?

Reply:
Well the `finally` actually contributes to the mistaken logic of the bug.

This code correctly implements what is specified in the documentation:

> Automatic acking (AckMode.AUTO)
> 
> A message is acked with GCP Pub/Sub if the adapter sent it to the channel and no exceptions were thrown.
> 
> If a RuntimeException is thrown while the message is processed, then the message is handled in the following way:
> 
> If an error channel is configured for the adapter (pubSubInboundChannelAdapater.getErrorChannel() != null) then the message will be acked and forwarded with the error to the error channel.
> 
> If no error channel is configured for the adapter, then the message is nacked.
> 
> Automatic acking OK (AckMode.AUTO_ACK)
> 
> A message is acked with GCP Pub/Sub if the adapter sent it to the channel and no exceptions were thrown. If a RuntimeException is thrown while the message is processed, then the message is neither acked / nor nacked.
> 
> This is useful when using the subscription’s ack deadline timeout as a retry delivery backoff mechanism.

https://github.com/spring-cloud/spring-cloud-gcp/blob/8e4fe4dcf56b51f6eba3cc8085dc5e9877fe7055/docs/src/main/asciidoc/spring-integration-pubsub.adoc

Reply:
I like the new way; it's cleaner and easier to analyze. Java `finally` behavior, especially where exceptions are concerned, can be tricky.